rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala

```diff
@@ -0,0 +1,292 @@
+package coop.rchain.rspace.history
+
+import cats.syntax.all._
+import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.history.History.KeyPath
+import coop.rchain.rspace.history.TestData._
+import coop.rchain.shared.Base16
+import coop.rchain.store.InMemoryKeyValueStore
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.{Assertion, FlatSpec, Matchers, OptionValues}
+import scodec.bits.ByteVector
+
+import java.nio.ByteBuffer
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.duration._
+import scala.util.Random
+
+class HistoryActionTests
+    extends FlatSpec
+    with Matchers
+    with OptionValues
+    with InMemoryHistoryTestBase {
+
+  "creating and read one record" should "works" in withEmptyHistory { emptyHistoryF =>
+    val data = insert(_zeros) :: Nil
+    for {
+      emptyHistory <- emptyHistoryF
+      newHistory   <- emptyHistory.process(data)
+      readValue    <- newHistory.read(_zeros)
+      _            = readValue shouldBe data.head.hash.bytes.some
+    } yield ()
+  }
+
+  "reset method of history" should "works" in withEmptyHistory { emptyHistoryF =>
+    val data = insert(_zeros) :: Nil
+    for {
+      emptyHistory    <- emptyHistoryF
+      newHistory      <- emptyHistory.process(data)
+      historyOneReset <- emptyHistory.reset(newHistory.root)
+      readValue       <- historyOneReset.read(_zeros)
+      _               = readValue shouldBe data.head.hash.bytes.some
+    } yield ()
+  }
+
+  "creating ten records" should "works" in withEmptyHistory { emptyHistoryF =>
+    val data = List.range(0, 10).map(zerosAnd).map(k => InsertAction(k, randomBlake))
+    for {
+      emptyHistory <- emptyHistoryF
+      newHistory   <- emptyHistory.process(data)
+      readValues   <- data.traverse(action => newHistory.read(action.key))
+      _            = readValues shouldBe data.map(_.hash.bytes.some)
+    } yield ()
+  }
+
+  "History" should "allow to store different length key records in different branches" in withEmptyHistory {
+    emptyHistoryF =>
+      val data = insert(hexKey("01")) :: insert(hexKey("02")) ::
+        insert(hexKey("0001")) :: insert(hexKey("0002")) :: Nil
+      for {
+        emptyHistory <- emptyHistoryF
+        newHistory   <- emptyHistory.process(data)
+        readValues   <- data.traverse(action => newHistory.read(action.key))
+        _            = readValues shouldBe data.map(_.hash.bytes.some)
+      } yield ()
+  }
+
+  // TODO: Don't works for MergingHistory
```

Comment:
What does this `Don't works for MergingHistory` mean?

Reply:
I don't know why, but MergingHistory does not pass this test. But for RadixHistory this test is okey.